### PR TITLE
New properties in NavigationSettings

### PR DESCRIPTION
**Pull request by S2Ler**
PassiveLocationManager, RouteController gives you a false sense that you can have various tile store configs for these controllers, but we have just a singleton Navigator. I found that different application parts depend on tile store config + Directions instances. While this is more or less ok using many instances of Directions, it is not ok to use different tile store configs. You need a heavy-weight cache handler to use tileset descriptors, so we shouldn't have more than one. And cache handlers depend on a tile store config.

So to communicate better to the developers that Navigation SDK should use one instance of tileStoreConfiguration and provide ability to change default Directions instance used across the SDK, I added new parameters to NavigationSettings and a new initialize method which can be used to override these parameters. This change is a breaking change and might require fixing docs/examples.

### CODE REVIEW

File: Sources/MapboxCoreNavigation/NavSDK.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import Foundation
+import MapboxDirections
+
+public struct NavSDK {
```
Comment (1ec5) [inline Sources/MapboxCoreNavigation/NavSDK.swift:4]: Consider combining this struct with NavigationSettings.

File: Sources/MapboxCoreNavigation/NavSDK.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    private static let lock: NSLock = .init()
+    private static var isInitialized: Bool = false
+    public static var directions: Directions!
+    public static var tileStoreConfiguration: TileStoreConfiguration!
+
+    public static func initialize(directions: Directions = .shared,
+                                  tileStoreConfiguration: TileStoreConfiguration = .default) {
```
Comment (1ec5) [inline Sources/MapboxCoreNavigation/NavSDK.swift:11]: Not too keen on an initializer that has to be called before everything else in the application. Developers would be prone to forgetting to call it, and overall it feels like a non-Cocoa way of setting up a library versus something more tightly scoped. If this code really has to run as soon as possible, then consider sticking an Objective-C `+initialize` method in CMapboxCoreNavigation.

File: Sources/MapboxCoreNavigation/NavigationService.swift [hunk 3]
```diff
@@ -151,7 +151,9 @@ public class MapboxNavigationService: NSObject, NavigationService {
     /**
      A reference to a MapboxDirections service. Used for rerouting.
      */
-    public var directions: Directions
+    public var directions: Directions {
+        NavSDK.directions
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/NavigationService.swift:155]: There is a case when you want to request a route from one endpoint and also use it for rerouting, but use a different endpoint for navigation tiles(in Navigator)

File: Sources/MapboxCoreNavigation/NavSDK.swift [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import Foundation
+import MapboxDirections
+
+public struct NavSDK {
+    private static let lock: NSLock = .init()
+    private static var isInitialized: Bool = false
+    public static var directions: Directions!
+    public static var tileStoreConfiguration: TileStoreConfiguration!
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/NavSDK.swift:8]: What if we use default values here and don't require calling `initialize` at the very start of the app?

File: Sources/MapboxCoreNavigation/PassiveLocationManager.swift [hunk 5]
```diff
@@ -40,7 +37,9 @@ open class PassiveLocationManager: NSObject {
     /**
      The directions service that allows the location manager to access road network data.
      */
-    public let directions: Directions
+    public var directions: Directions {
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/PassiveLocationManager.swift:40]: Do we need such variables after these changes? I would remove them or make them private

File: Sources/MapboxCoreNavigation/NavSDK.swift (hunk 2 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavSDK.swift:11]: As per discussion, this will become an advance method with fallback to defaults.

File: Sources/MapboxCoreNavigation/NavigationService.swift (hunk 3 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationService.swift:155]: I decided to keep previous implementation and just default to the NavSDK.directions instead of simply .shared.

File: Sources/MapboxCoreNavigation/NavSDK.swift (hunk 4 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavSDK.swift:8]: Yep, will do, thanks for suggestion. It will be the best of two worlds.

File: Sources/MapboxCoreNavigation/PassiveLocationManager.swift (hunk 5 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/PassiveLocationManager.swift:40]: I reverted the change

File: Sources/MapboxCoreNavigation/NavSDK.swift (hunk 1 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavSDK.swift]: done, please review the latest revision.

File: Sources/MapboxCoreNavigation/NavSDK.swift (hunk 2 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavSDK.swift]: This is no longer a required call. Please review the latest revision.

File: Sources/MapboxCoreNavigation/NavigationService.swift (hunk 3 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationService.swift]: Done, please review the latest revision.

File: Sources/MapboxCoreNavigation/NavSDK.swift (hunk 4 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavSDK.swift]: Done, please review the latest revision.

File: Sources/MapboxCoreNavigation/PassiveLocationManager.swift (hunk 5 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/PassiveLocationManager.swift]: Done, please review the latest revision.

File: Sources/MapboxCoreNavigation/NavigationSettings.swift [hunk 6]
```diff
@@ -23,6 +24,47 @@ public class NavigationSettings {
             }
         }
     }
+
+    private let lock: NSLock = .init()
+    private var _directions: Directions?
+    private var _tileStoreConfiguration: TileStoreConfiguration?
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/NavigationSettings.swift:30]: I would suggest creating a struct with these two properties and store that struct. This reduces code duplication in accessors.

File: Sources/MapboxCoreNavigation/NavigationSettings.swift [hunk 7]
```diff
@@ -73,17 +115,6 @@ public class NavigationSettings {
      */
     public static let shared = NavigationSettings()
     
-    /// Returns a reflection of this class excluding the `properties` variable.
-    lazy var properties: [Mirror.Child] = {
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/NavigationSettings.swift:77]: Why has this been removed?

Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationSettings.swift:77]: It is not used anywhere. Do I miss something?

File: Sources/MapboxCoreNavigation/NavigationSettings.swift (hunk 6 above)
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationSettings.swift]: Done, please review the latest revision.

**Comment by S2Ler** [comment]
Changelog is coming

**Review by 1ec5 — APPROVED** [review]
I’m really liking how much simpler the code is becoming this way. I wonder if there’s a way to make the lifecycle clearer to developers through more extensive use of UserDefaults, as an optional alternative to calling `initialize`. But apart from some documentation-related nits, I don’t see any red flags with this approach.

File: Sources/MapboxCoreNavigation/NavigationSettings.swift [hunk 8]
```diff
@@ -23,6 +27,76 @@ public class NavigationSettings {
...
+     Not doing so will lead to undefined behavior.
+
+     - Parameters:
+       - directions: Default `Directions` instance. Some types allow you to customize the directions instance and
+     fall back to the `NavigationSettings.directions` by default.
+       - tileStoreConfiguration: Options for configuring how map and navigation tiles are stored on the device. See
+     `TileStoreConfiguration` for more details.
+     */
+    public func initialize(directions: Directions,
+                           tileStoreConfiguration: TileStoreConfiguration) {
```
[the remark this answers is not in the document]
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationSettings.swift:91]: I will explore this approach, thanks for suggestion, I like it

File: Sources/MapboxCoreNavigation/NavigationSettings.swift [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
 
 /**
- A wrapper for the `UserDefaults` class for navigation-specific settings.
- 
- Properties are prefixed before they are stored in `UserDefaults.standard`.
- 
- To specify criteria when calculating routes, use the `NavigationRouteOptions` class. To customize the user experience during a particular turn-by-turn navigation session, use the `NavigationOptions` class when initializing a `NavigationViewController`.
+ Global settings that are used across the SDK for altering navigation behavior.
+
+ To specify criteria when calculating routes, use the `NavigationRouteOptions` class.
```
[the remark this answers is not in the document]
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationSettings.swift:7]: Done, please review the latest revision.

File: Sources/MapboxCoreNavigation/NavigationSettings.swift [hunk 10]
```diff
@@ -23,6 +27,76 @@ public class NavigationSettings {
...
+        state.tileStoreConfiguration
+    }
+
+    /**
+     Initializes the settings with custom instances of globally used types.
+
+     If you don't provide custom values, they will be initialized with the defaults.
+
+     - important: If you want to use this method, you should call it before you start using Navigation SDK.
+     Not doing so will lead to undefined behavior.
```
[the remark this answers is not in the document]
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationSettings.swift:82]: Please help me with this. I've changed the wording a little bit but not sure how to make it more precise without being too wordy and unnecessary specific.

File: Sources/MapboxCoreNavigation/NavigationSettings.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationSettings.swift:91]: Given that it is an improvement, I will create a tail work ticket for this suggestion.

File: CHANGELOG.md [hunk 11]
```diff
@@ -139,7 +138,15 @@
...
+* Removed `NavigationOptions.tileStoreConfiguration` in favor of `NavigationSettings.tileStoreConfiguration`. ([#3215](https://github.com/mapbox/mapbox-navigation-ios/pull/3215))
+* Removed `tileStoreConfiguration` parameter `PredictiveCacheManager.init(predictiveCacheOptions:styleSourcePaths:)` in favor of `NavigationSettings.tileStoreConfiguration`. ([#3215](https://github.com/mapbox/mapbox-navigation-ios/pull/3215))
+* `NavigationViewController.mapTileStore` is now obtained from `NavigationSettings.tileStoreConfiguration`. ([#3215](https://github.com/mapbox/mapbox-navigation-ios/pull/3215))
+* Removed `tileStoreConfiguration` parameter from `NavigationMapView.enablePredictiveCaching(options:)` in favor of `NavigationSettings.tileStoreConfiguration`. ([#3215](https://github.com/mapbox/mapbox-navigation-ios/pull/3215))
```
[the remark this answers is not in the document]
Comment (S2Ler) [reply CHANGELOG.md:152]: Where would I put these changelong entries?

File: Sources/MapboxCoreNavigation/NavigationSettings.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (S2Ler) [reply Sources/MapboxCoreNavigation/NavigationSettings.swift:93]: #3259

File: CHANGELOG.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (1ec5) [reply CHANGELOG.md:159]: #3376 removed the blurbs from the running changelog, but in the future, you can add intra-release changes to the PR description and/or a draft release’s release notes.
